python/ray/serve/_private/serve_request_response.py

```diff
@@ -0,0 +1,152 @@
+from abc import ABC
+import grpc
+import logging
+import pickle
+
+from starlette.types import Receive, Scope, Send
+from typing import Any, Callable, List, Generator, Optional, Tuple
+
+from ray.actor import ActorHandle
+from ray.serve._private.constants import SERVE_LOGGER_NAME
+from ray.serve._private.utils import DEFAULT
+from ray.serve._private.common import gRPCRequest, StreamingHTTPRequest
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+class ServeRequest(ABC):
+    """Base ServeRequest class to use in the common interface among proxies"""
+
+    pass
+
+
+class ASGIServeRequest(ServeRequest):
+    """ServeRequest implementation to wrap ASGI scope, receive, and send."""
+
+    def __init__(self, scope: Scope, receive: Receive, send: Send):
+        self.scope = scope
+        self.receive = receive
+        self.send = send
+
+    @property
+    def request_type(self) -> str:
+        return self.scope.get("type", "")
+
+    @property
+    def client(self) -> str:
+        return self.scope.get("client", "")
+
+    @property
+    def method(self) -> str:
+        return self.scope.get("method", "websocket").upper()
+
+    @property
+    def root_path(self) -> str:
+        return self.scope.get("root_path", "")
+
+    @property
+    def route_path(self) -> str:
+        return self.scope.get("path", "")[len(self.root_path) :]
```

Comment:
is it a issue when `self.scope.get("path", "")` returns ""?

nit: [len(self.root_path):]